### DISCUSSION TITLE
zrepl over unstable network

### DISCUSSION THREAD

**Discussion by TPXP** [zrepl/zrepl#834] [Q&A]
Hello,

I'm using zrepl to replicate data between two homemade NAS powered by Raspberry Pis (with Wireguard to encrypt data over the Internet). Those systems are not very powerful and, when put under stress with a large data transfer (Wireguard encryption is quite CPU-intensive), they tend to perform weirdly and sometimes their internet connection breaks briefly.

```
[5609024.863396] bcmgenet fd580000.ethernet end0: Link is Down
[5609027.936814] bcmgenet fd580000.ethernet end0: Link is Up - 1Gbps/Full - flow control off<
```

When this happens in the middle of a zrepl transfer, logs look like this
* Sender:
```
Aug 25 09:32:13 rpi4-par zrepl[945]: 2024-08-25T09:32:13Z [ERRO][job][repl][<some weird stuff>]: receive request failed (might also be error on sender) err="stream: read error: read tcp 10.0.1.4:48194->192.168.10.2:8888: i/o timeout" errType="*stream.ReadStreamError" rr="Filesystem:\"vol/pool\" To:{Name:\"20240825\" Guid:644720186733263380 CreateTXG:3398448 Creation:\"2024-08-25T09:27:58Z\"} ClearResumeToken:true ReplicationConfig:{protection:{Initial:GuaranteeResumability Incremental:GuaranteeResumability}}" filesystem="data_par/data_par"
Aug 25 09:32:13 rpi4-par zrepl[945]: 2024-08-25T09:32:13Z [ERRO][job][repl][<some weird stuff>]: most recent error in this attempt err="stream: read error: read tcp 10.0.1.4:48194->192.168.10.2:8888: i/o timeout" attempt_number="0"
Aug 25 09:32:13 rpi4-par zrepl[945]: 2024-08-25T09:32:13Z [ERRO][job][repl][<some weird stuff>]: temporary connectivity-related error identified, start waiting for reconnect deadline="2024-08-25 09:42:13.191851221 +0000 UTC m=+5610056.101978219" attempt_number="0"
Aug 25 09:32:13 rpi4-par zrepl[945]: 2024-08-25T09:32:13Z [ERRO][job][repl][<some weird stuff>]: reconnect successful attempt_number="0"
```
* Receiver:
```
Aug 25 09:32:13 rpi4-nce zrepl[722]: 2024-08-25T09:32:13Z [ERRO][sink][endpoint][<some weird stuff>]: zfs receive failed err="zfs exited with error: signal: killed\nstderr:\n" proto_fs="vol/pool" local_fs="vol/pool/zzz/test" opts="zfs.RecvOptions{RollbackAndForceRecv:false, SavePartialRecvState:true, InheritProperties:[]property.Property{}, OverrideProperties:map[property.Property]string{}}"
Aug 25 09:32:13 rpi4-nce zrepl[722]: 2024-08-25T09:32:13Z [ERRO][sink][rpc.data][<some weird stuff>]: cannot write response header err="frameconn: shutting down"
```

On the sender, `zrepl status` gives a scary error `filesystem-error` (which I spontaneously interpret as "faulty drive")
<img width="832" alt="image" src="https://github.com/user-attachments/assets/9ea36dfd-3184-4971-abd6-7834e59d7750">

I recon my network is bad, but networking issues are hard to fix so I'd love to be able to configure zrepl's retries behaviour for these situations to be handled more gracefully. I couldn't find a section about retries in the documentation, did I miss something?

Thanks in advance for your help, zrepl is awesome!

**Comment by problame** [comment]
Sorry for the late reply.

While the error message you're seeing (`filesystem-error`) looks scary, you correctly identified that this is just due to your connectivity issues.

The log says

```
temporary connectivity-related error identified, start waiting for reconnect
```

and in your screenshot you see `Attempt #2`.

zrepl _does_ automatically retry on connectivity related errors - the `#2` is a witness of that.

So, let's first figure out: is this issue about noisy logs or does the reconnect not work?

---

There is no configuration option for retries right now.
The classification code for retryable vs permanent errors is here:

https://github.com/zrepl/zrepl/blob/908807bd59d8fa76bf4db51b4ca64466fba65ae1/internal/replication/driver/replication_driver.go#L805-L818

**Comment by TPXP** [comment]
Thanks for your reply, no worries for the delay!

I didn't pay much attention to `Attempt #2`, since there's nothing next to it. What catches my attention is the failed attempt and I didn't know zrepl had a retry mechanism (which is a good idea - thanks for implementing it :) ), so I restart the job with `zrepl signal wakeup` when this happens. If zrepl is just waiting a little bit before resuming a new sync, I think the status output could clarify if - maybe say something like "waiting xx seconds..."?

**Comment by problame** [comment]
>  I think the status output could clarify if - maybe say something like "waiting xx seconds..."?

Hmm, we do actually print something like that, perhaps too technical / gibberish?

https://github.com/zrepl/zrepl/blob/908807bd59d8fa76bf4db51b4ca64466fba65ae1/internal/client/status/viewmodel/render.go#L296-L315

Happy to accept a PR that clarifies the language; please provide a screenshot as part of the PR that illustrates the improvement!

**Comment by problame** [comment]
Moving this into Q&A since it turned out to be not a functional issue .

Tracking the evident UX issue here, patches are very welcome :) 
https://github.com/zrepl/zrepl/issues/833
